packages/iris-grid/src/sidebar/VisibilityOrderingBuilder.tsx

```diff
@@ -675,77 +617,333 @@ class VisibilityOrderingBuilder extends Component<
+            deleteNames.add(parent.name);
+            parent.parent?.removeChildren([parent.name]);
+            parent = parent.parent;
+          }
+
+          // Delete all groups that are now empty
+          newGroups = newGroups.filter(({ name }) => !deleteNames.has(name));
+        }
       }
 
-      wasMovable = isMovable;
+      if (toGroup != null) {
+        // Moved into a group
+        toGroup.addChildren([from.id.toString()]);
+      }
+
+      model.setColumnHeaderGroups(newGroups);
+      onColumnHeaderGroupChanged(newGroups);
+    }
+
+    const fromIndex = from.data.visibleIndex;
+
+    let toIndex = Array.isArray(to.data.visibleIndex)
+      ? to.data.visibleIndex[0]
+      : to.data.visibleIndex;
+
+    // Dropping as first item in a group
+    // Need to adjust visible index if dragging from before this group or it is off by 1
+    if (
+      to.children.length > 0 &&
+      (Array.isArray(fromIndex) ? fromIndex[0] : fromIndex) < toIndex
+    ) {
+      toIndex -= 1;
     }
 
-    return elements;
+    onMovedColumnsChanged(
+      Array.isArray(fromIndex)
+        ? GridUtils.moveRange(fromIndex, toIndex, movedColumns, true)
+        : GridUtils.moveItem(fromIndex, toIndex, movedColumns)
+    );
   }
 
-  getFirstMovableIndex = memoize((model: IrisGridModel, columns: Column[]) => {
-    for (let i = 0; i < columns.length; i += 1) {
-      if (model.isColumnMovable(i)) {
-        return i;
+  handleGroupColorChange = throttle(
+    (group: ColumnHeaderGroup, color: string | undefined): void => {
+      const { model, onColumnHeaderGroupChanged } = this.props;
+      const columnGroups = model.getColumnHeaderGroups();
+      const newGroups = [...columnGroups];
+      const newGroup = newGroups.find(({ name }) => name === group.name);
+      if (!newGroup) {
+        throw new Error(
+          `Changed the color group that does not exist: ${group.name}`
+        );
       }
+
+      newGroup.color = color;
+
+      model.setColumnHeaderGroups(newGroups);
+      onColumnHeaderGroupChanged(newGroups);
+    },
+    250
+  );
+
+  handleGroupDelete(group: ColumnHeaderGroup): void {
+    const { model, onColumnHeaderGroupChanged } = this.props;
+    const columnGroups = model.getColumnHeaderGroups();
+    const newGroups = columnGroups.filter(g => g.name !== group.name);
+    const newParent = newGroups.find(g => g.name === group.parent?.name);
+    if (newParent !== undefined) {
+      newParent.addChildren(group.children);
+      newParent.removeChildren([group.name]);
     }
+    model.setColumnHeaderGroups(newGroups);
+    onColumnHeaderGroupChanged(newGroups);
+  }
+
+  handleGroupCreate(childIndexes: number[]): void {
+    const { model, onColumnHeaderGroupChanged } = this.props;
+    const columnGroupMap = model.getColumnHeaderGroupMap();
+    const newGroups = [...columnGroupMap.values()];
+
+    // TODO: Change to name based so you can seed with groups
+    // TODO: Remove children from their parents
```

Comment:
Some TODOs. Noticed some errors:
- Creating a group when a group named `NewGroup` already exists
- Creating a group with child elements selected
- Creating a group with some top level and child elements selected

Reply:
These should all be fixed now